File: binding/include/network/libqhttpheaders.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qhttpheaders.html
pub const qhttpheaders = struct {

    /// New constructs a new QHttpHeaders object.
    pub fn New() QtC.QHttpHeaders {
        return qtc.QHttpHeaders_new();
    }


    /// New2 constructs a new QHttpHeaders object.
    pub fn New2(other: ?*anyopaque) QtC.QHttpHeaders {
        return qtc.QHttpHeaders_new2(@ptrCast(other));
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QHttpHeaders_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QHttpHeaders_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn Contains2(self: ?*anyopaque, name: i32) bool {
        return qtc.QHttpHeaders_Contains2(@ptrCast(self), @intCast(name));
    }

    pub fn Clear(self: ?*anyopaque, ) void {
        qtc.QHttpHeaders_Clear(@ptrCast(self));
    }

    pub fn RemoveAll2(self: ?*anyopaque, name: i32) void {
        qtc.QHttpHeaders_RemoveAll2(@ptrCast(self), @intCast(name));
    }

    pub fn RemoveAt(self: ?*anyopaque, i: i64) void {
        qtc.QHttpHeaders_RemoveAt(@ptrCast(self), i);
    }

    pub fn Value2(self: ?*anyopaque, name: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QHttpHeaders_Value2(@ptrCast(self), @intCast(name));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qhttpheaders.Value2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Values2(self: ?*anyopaque, name: i32) []const u8 {
        return qtc.QHttpHeaders_Values2(@ptrCast(self), @intCast(name));
    }

    pub fn ValueAt(self: ?*anyopaque, i: i64, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QHttpHeaders_ValueAt(@ptrCast(self), i);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qhttpheaders.ValueAt: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn CombinedValue2(self: ?*anyopaque, name: i32, allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QHttpHeaders_CombinedValue2(@ptrCast(self), @intCast(name));
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qhttpheaders.CombinedValue2: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn Size(self: ?*anyopaque, ) i64 {
        return qtc.QHttpHeaders_Size(@ptrCast(self));
    }

    pub fn Reserve(self: ?*anyopaque, size: i64) void {
        qtc.QHttpHeaders_Reserve(@ptrCast(self), size);
    }

    pub fn IsEmpty(self: ?*anyopaque, ) bool {
        return qtc.QHttpHeaders_IsEmpty(@ptrCast(self));
    }

    pub fn WellKnownHeaderName(name: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QHttpHeaders_WellKnownHeaderName(@intCast(name));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qhttpheaders.WellKnownHeaderName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn FromListOfPairs(headers: []const u8) QtC.QHttpHeaders {
        return qtc.QHttpHeaders_FromListOfPairs(@ptrCast(headers));
    }

    pub fn FromMultiMap(headers: *const QMultiMap<QByteArray, QByteArray>) QtC.QHttpHeaders {
        return qtc.QHttpHeaders_FromMultiMap(@ptrCast(headers));
    }

    pub fn FromMultiHash(headers: *const QMultiHash<QByteArray, QByteArray>) QtC.QHttpHeaders {
        return qtc.QHttpHeaders_FromMultiHash(@ptrCast(headers));
    }

    pub fn ToListOfPairs(self: ?*anyopaque, ) []const u8 {
        return qtc.QHttpHeaders_ToListOfPairs(@ptrCast(self));
    }

    pub fn ToMultiMap(self: ?*anyopaque, ) QMultiMap<QByteArray, QByteArray> {
        return qtc.QHttpHeaders_ToMultiMap(@ptrCast(self));
    }

    pub fn ToMultiHash(self: ?*anyopaque, ) QMultiHash<QByteArray, QByteArray> {
        return qtc.QHttpHeaders_ToMultiHash(@ptrCast(self));
    }

    pub fn Value23(self: ?*anyopaque, name: i32, defaultValue: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QHttpHeaders_Value23(@ptrCast(self), @intCast(name), @ptrCast(defaultValue));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qhttpheaders.Value23: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QHttpHeaders_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/httpheaders.html#types
pub const enums = struct {
    pub const WellKnownHeader = enum {
        pub const AIM: i32 = 0;
        pub const Accept: i32 = 1;
        pub const AcceptAdditions: i32 = 2;
        pub const AcceptCH: i32 = 3;
        pub const AcceptDatetime: i32 = 4;
        pub const AcceptEncoding: i32 = 5;
        pub const AcceptFeatures: i32 = 6;
        pub const AcceptLanguage: i32 = 7;
        pub const AcceptPatch: i32 = 8;
        pub const AcceptPost: i32 = 9;
        pub const AcceptRanges: i32 = 10;
        pub const AcceptSignature: i32 = 11;
        pub const AccessControlAllowCredentials: i32 = 12;
        pub const AccessControlAllowHeaders: i32 = 13;
        pub const AccessControlAllowMethods: i32 = 14;
        pub const AccessControlAllowOrigin: i32 = 15;
        pub const AccessControlExposeHeaders: i32 = 16;
        pub const AccessControlMaxAge: i32 = 17;
        pub const AccessControlRequestHeaders: i32 = 18;
        pub const AccessControlRequestMethod: i32 = 19;
        pub const Age: i32 = 20;
        pub const Allow: i32 = 21;
        pub const ALPN: i32 = 22;
        pub const AltSvc: i32 = 23;
        pub const AltUsed: i32 = 24;
        pub const Alternates: i32 = 25;
        pub const ApplyToRedirectRef: i32 = 26;
        pub const AuthenticationControl: i32 = 27;
        pub const AuthenticationInfo: i32 = 28;
        pub const Authorization: i32 = 29;
        pub const CacheControl: i32 = 30;
        pub const CacheStatus: i32 = 31;
        pub const CalManagedID: i32 = 32;
        pub const CalDAVTimezones: i32 = 33;
        pub const CapsuleProtocol: i32 = 34;
        pub const CDNCacheControl: i32 = 35;
        pub const CDNLoop: i32 = 36;
        pub const CertNotAfter: i32 = 37;
        pub const CertNotBefore: i32 = 38;
        pub const ClearSiteData: i32 = 39;
        pub const ClientCert: i32 = 40;
        pub const ClientCertChain: i32 = 41;
        pub const Close: i32 = 42;
        pub const Connection: i32 = 43;
        pub const ContentDigest: i32 = 44;
        pub const ContentDisposition: i32 = 45;
        pub const ContentEncoding: i32 = 46;
        pub const ContentID: i32 = 47;
        pub const ContentLanguage: i32 = 48;
        pub const ContentLength: i32 = 49;
        pub const ContentLocation: i32 = 50;
        pub const ContentRange: i32 = 51;
        pub const ContentSecurityPolicy: i32 = 52;
        pub const ContentSecurityPolicyReportOnly: i32 = 53;
        pub const ContentType: i32 = 54;
        pub const Cookie: i32 = 55;
        pub const CrossOriginEmbedderPolicy: i32 = 56;
        pub const CrossOriginEmbedderPolicyReportOnly: i32 = 57;
        pub const CrossOriginOpenerPolicy: i32 = 58;
        pub const CrossOriginOpenerPolicyReportOnly: i32 = 59;
        pub const CrossOriginResourcePolicy: i32 = 60;
        pub const DASL: i32 = 61;
        pub const Date: i32 = 62;
        pub const DAV: i32 = 63;
        pub const DeltaBase: i32 = 64;
        pub const Depth: i32 = 65;
        pub const Destination: i32 = 66;
        pub const DifferentialID: i32 = 67;
        pub const DPoP: i32 = 68;
        pub const DPoPNonce: i32 = 69;
        pub const EarlyData: i32 = 70;
        pub const ETag: i32 = 71;
        pub const Expect: i32 = 72;
        pub const ExpectCT: i32 = 73;
        pub const Expires: i32 = 74;
        pub const Forwarded: i32 = 75;
        pub const From: i32 = 76;
        pub const Hobareg: i32 = 77;
        pub const Host: i32 = 78;
        pub const If: i32 = 79;
        pub const IfMatch: i32 = 80;
        pub const IfModifiedSince: i32 = 81;
        pub const IfNoneMatch: i32 = 82;
        pub const IfRange: i32 = 83;
        pub const IfScheduleTagMatch: i32 = 84;
        pub const IfUnmodifiedSince: i32 = 85;
        pub const IM: i32 = 86;
        pub const IncludeReferredTokenBindingID: i32 = 87;
        pub const KeepAlive: i32 = 88;
        pub const Label: i32 = 89;
        pub const LastEventID: i32 = 90;
        pub const LastModified: i32 = 91;
        pub const Link: i32 = 92;
        pub const Location: i32 = 93;
        pub const LockToken: i32 = 94;
        pub const MaxForwards: i32 = 95;
        pub const MementoDatetime: i32 = 96;
        pub const Meter: i32 = 97;
        pub const MIMEVersion: i32 = 98;
        pub const Negotiate: i32 = 99;
        pub const NEL: i32 = 100;
        pub const ODataEntityId: i32 = 101;
        pub const ODataIsolation: i32 = 102;
        pub const ODataMaxVersion: i32 = 103;
        pub const ODataVersion: i32 = 104;
        pub const OptionalWWWAuthenticate: i32 = 105;
        pub const OrderingType: i32 = 106;
        pub const Origin: i32 = 107;
        pub const OriginAgentCluster: i32 = 108;
        pub const OSCORE: i32 = 109;
        pub const OSLCCoreVersion: i32 = 110;
        pub const Overwrite: i32 = 111;
        pub const PingFrom: i32 = 112;
        pub const PingTo: i32 = 113;
        pub const Position: i32 = 114;
        pub const Prefer: i32 = 115;
        pub const PreferenceApplied: i32 = 116;
        pub const Priority: i32 = 117;
        pub const ProxyAuthenticate: i32 = 118;
        pub const ProxyAuthenticationInfo: i32 = 119;
        pub const ProxyAuthorization: i32 = 120;
        pub const ProxyStatus: i32 = 121;
        pub const PublicKeyPins: i32 = 122;
        pub const PublicKeyPinsReportOnly: i32 = 123;
        pub const Range: i32 = 124;
        pub const RedirectRef: i32 = 125;
        pub const Referer: i32 = 126;
        pub const Refresh: i32 = 127;
        pub const ReplayNonce: i32 = 128;
        pub const ReprDigest: i32 = 129;
        pub const RetryAfter: i32 = 130;
        pub const ScheduleReply: i32 = 131;
        pub const ScheduleTag: i32 = 132;
        pub const SecPurpose: i32 = 133;
        pub const SecTokenBinding: i32 = 134;
        pub const SecWebSocketAccept: i32 = 135;
        pub const SecWebSocketExtensions: i32 = 136;
        pub const SecWebSocketKey: i32 = 137;
        pub const SecWebSocketProtocol: i32 = 138;
        pub const SecWebSocketVersion: i32 = 139;
        pub const Server: i32 = 140;
        pub const ServerTiming: i32 = 141;
        pub const SetCookie: i32 = 142;
        pub const Signature: i32 = 143;
        pub const SignatureInput: i32 = 144;
        pub const SLUG: i32 = 145;
        pub const SoapAction: i32 = 146;
        pub const StatusURI: i32 = 147;
        pub const StrictTransportSecurity: i32 = 148;
        pub const Sunset: i32 = 149;
        pub const SurrogateCapability: i32 = 150;
        pub const SurrogateControl: i32 = 151;
        pub const TCN: i32 = 152;
        pub const TE: i32 = 153;
        pub const Timeout: i32 = 154;
        pub const Topic: i32 = 155;
        pub const Traceparent: i32 = 156;
        pub const Tracestate: i32 = 157;
        pub const Trailer: i32 = 158;
        pub const TransferEncoding: i32 = 159;
        pub const TTL: i32 = 160;
        pub const Upgrade: i32 = 161;
        pub const Urgency: i32 = 162;
        pub const UserAgent: i32 = 163;
        pub const VariantVary: i32 = 164;
        pub const Vary: i32 = 165;
        pub const Via: i32 = 166;
        pub const WantContentDigest: i32 = 167;
        pub const WantReprDigest: i32 = 168;
        pub const WWWAuthenticate: i32 = 169;
        pub const XContentTypeOptions: i32 = 170;
        pub const XFrameOptions: i32 = 171;
        pub const AcceptCharset: i32 = 172;
        pub const CPEPInfo: i32 = 173;
        pub const Pragma: i32 = 174;
        pub const ProtocolInfo: i32 = 175;
        pub const ProtocolQuery: i32 = 176;
    };

};

File: binding/src/network/libqhttpheaders.cpp
#include <QByteArray>
#include <QByteArrayView>
#include <QHttpHeaders>
#include <qhttpheaders.h>
#include "libqhttpheaders.h"
#include "libqhttpheaders.hxx"

QHttpHeaders* QHttpHeaders_new() {
	 return new QHttpHeaders();
}

QHttpHeaders* QHttpHeaders_new2(const QHttpHeaders* other) {
	 return new QHttpHeaders(*other);
}

void QHttpHeaders_OperatorAssign(QHttpHeaders* self, const QHttpHeaders* other) {
	self->operator=(*other);
}

void QHttpHeaders_Swap(QHttpHeaders* self, QHttpHeaders* other) {
	self->swap(*other);
}

bool QHttpHeaders_Contains2(const QHttpHeaders* self, int name) {
	return self->contains(static_cast<QHttpHeaders::WellKnownHeader>(name));
}

void QHttpHeaders_Clear(QHttpHeaders* self) {
	self->clear();
}

void QHttpHeaders_RemoveAll2(QHttpHeaders* self, int name) {
	self->removeAll(static_cast<QHttpHeaders::WellKnownHeader>(name));
}

void QHttpHeaders_RemoveAt(QHttpHeaders* self, ptrdiff_t i) {
	self->removeAt(i);
}

QByteArrayView* QHttpHeaders_Value2(const QHttpHeaders* self, int name) {
	return new QByteArrayView(self->value(static_cast<QHttpHeaders::WellKnownHeader>(name)));
}

libqt_list QHttpHeaders_Values2(const QHttpHeaders* self, int name) {
	auto _ret = self->values(static_cast<QHttpHeaders::WellKnownHeader>(name));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

QByteArrayView* QHttpHeaders_ValueAt(const QHttpHeaders* self, ptrdiff_t i) {
	return new QByteArrayView(self->valueAt(i));
}

libqt_string QHttpHeaders_CombinedValue2(const QHttpHeaders* self, int name) {
	QByteArray _qb = self->combinedValue(static_cast<QHttpHeaders::WellKnownHeader>(name));
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

ptrdiff_t QHttpHeaders_Size(const QHttpHeaders* self) {
	return self->size();
}

void QHttpHeaders_Reserve(QHttpHeaders* self, ptrdiff_t size) {
	self->reserve(size);
}

bool QHttpHeaders_IsEmpty(const QHttpHeaders* self) {
	return self->isEmpty();
}

QByteArrayView* QHttpHeaders_WellKnownHeaderName(int name) {
	return new QByteArrayView(QHttpHeaders::wellKnownHeaderName(static_cast<QHttpHeaders::WellKnownHeader>(name)));
}

QHttpHeaders* QHttpHeaders_FromListOfPairs(const libqt_list headers) {
	return new QHttpHeaders(QHttpHeaders::fromListOfPairs(QList<QPair<QByteArray, QByteArray>>()));
}

QHttpHeaders* QHttpHeaders_FromMultiMap(const libqt_map headers) {
	return new QHttpHeaders(QHttpHeaders::fromMultiMap(QMultiMap<QByteArray, QByteArray>()));
}

QHttpHeaders* QHttpHeaders_FromMultiHash(const libqt_map headers) {
	return new QHttpHeaders(QHttpHeaders::fromMultiHash(QMultiHash<QByteArray, QByteArray>()));
}

libqt_list QHttpHeaders_ToListOfPairs(const QHttpHeaders* self) {
	auto _ret = self->toListOfPairs();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

libqt_map QHttpHeaders_ToMultiMap(const QHttpHeaders* self) {
	auto _ret = self->toMultiMap();
	libqt_map _map;
	_map.len = _ret.size();
	_map.keys = nullptr;
	_map.values = nullptr;
	return _map;
}

libqt_map QHttpHeaders_ToMultiHash(const QHttpHeaders* self) {
	auto _ret = self->toMultiHash();
	libqt_map _map;
	_map.len = _ret.size();
	_map.keys = nullptr;
	_map.values = nullptr;
	return _map;
}

QByteArrayView* QHttpHeaders_Value23(const QHttpHeaders* self, int name, QByteArrayView* defaultValue) {
	return new QByteArrayView(self->value(static_cast<QHttpHeaders::WellKnownHeader>(name), *defaultValue));
}

void QHttpHeaders_Delete(QHttpHeaders* self) {
    delete self;
}

